test/core/iomgr/endpoint_tests.cc

```diff
@@ -134,8 +135,12 @@ static void read_and_write_test_read_handler(void* data,
   struct read_and_write_test_state* state =
       static_cast<struct read_and_write_test_state*>(data);
 
-  state->bytes_read += count_slices(
-      state->incoming.slices, state->incoming.count, &state->current_read_data);
+  if (error.ok()) {
```

Comment:
Is this extra error check necessary? Seems weird to check `error.ok()` 3 times in a 15 line function.

Reply:
The test flakes without this checks. Sometimes the connection can get closed while the read callback is waiting for bytes to be read. If the socket returns -1 with no bytes read, then the status will be set to not ok. We should not attempt to call count_slices and read the slice buffer in this case. Previously this didn't happen because the test assumed that an ExecCtx::Flush will execute all pending callbacks and no read callback will be in wait while the socket gets closed.

Reply:
Here's a version that checks the error once:

```
static void read_and_write_test_read_handler_read_done(
    read_and_write_test_state* state) {
  gpr_log(GPR_DEBUG, "Read handler done");
  gpr_mu_lock(g_mu);
  GRPC_LOG_IF_ERROR("pollset_kick", grpc_pollset_kick(g_pollset, nullptr));
  gpr_mu_unlock(g_mu);
}

static void read_and_write_test_read_handler(void* data,
                                             grpc_error_handle error) {
  struct read_and_write_test_state* state =
      static_cast<struct read_and_write_test_state*>(data);
  if (!error.ok()) {
    state->read_done = 1;
    read_and_write_test_read_handler_read_done();
    return;
  }
  state->bytes_read += count_slices(
      state->incoming.slices, state->incoming.count, &state->current_read_data);
  if (state->bytes_read == state->target_bytes) {
    state->read_done = 2;
    read_and_write_test_read_handler_read_done();
    return;
  }
  // We perform many reads one after another. If grpc_endpoint_read and the
  // read_handler are both run inline, we might end up growing the stack
  // beyond the limit. Schedule the read on ExecCtx to avoid this.
  grpc_core::ExecCtx::Run(DEBUG_LOCATION, &state->read_scheduler,
                          absl::OkStatus());
}
```

Without thread annotations, it isn't super clear to me what needs to be guarded by g_mu, so the state->read_done value may need to move into the helper method (but I don't think so).

Reply:
I altered it to this format. I think read_done needs to be set under a mutex so its done inside the handler now.